xds/src/test/java/io/grpc/xds/XdsServerTestHelper.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2020 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.verify;
+
+import io.grpc.xds.internal.sds.XdsServerBuilder;
+import java.io.IOException;
+import java.net.ServerSocket;
+import org.mockito.ArgumentCaptor;
+
+/**
+ * Helper methods related to {@link XdsServerBuilder} and related classes.
+ */
+class XdsServerTestHelper {
+
+  static XdsClient.ListenerWatcher startAndGetWatcher(
+      XdsClientWrapperForServerSds xdsClientWrapperForServerSds,
+      XdsClient mockXdsClient,
+      int port) {
+    xdsClientWrapperForServerSds.start(mockXdsClient);
+    ArgumentCaptor<XdsClient.ListenerWatcher> listenerWatcherCaptor = ArgumentCaptor.forClass(null);
+    verify(mockXdsClient).watchListenerData(eq(port), listenerWatcherCaptor.capture());
+    return listenerWatcherCaptor.getValue();
+  }
+
+  static void generateListenerUpdate(
```

Comment:
I don't understand this method detail from it's name and args. Why there are two sets of port and context? Why there are two ports args but they are always the same value in all usages?

Reply:
So this is due to the complexity of the [Listener](https://github.com/envoyproxy/envoy/blob/master/api/envoy/config/listener/v3/listener.proto#L40) proto. This is really defined for the Envoy proxy which has a [physical address](https://github.com/envoyproxy/envoy/blob/master/api/envoy/config/listener/v3/listener.proto#L108) for the listener to listen on. And then the addresses in the filter chain which consist of [prefix_range](https://github.com/envoyproxy/envoy/blob/master/api/envoy/config/listener/v3/listener_components.proto#L96) and the [destination port](https://github.com/envoyproxy/envoy/blob/master/api/envoy/config/listener/v3/listener_components.proto#L96). Until we have the control plane (TD) send us the listener response especially for gRPC that has a single Listener object containing a single filter-chain with appropriate IP+port values I had to add support for processing and selecting the correct filter chain for the requested port. So the test needed to simulate those responses. Hope that answers the question. I'll add comments to this method to explain the need for those params.

Reply:
done and also cleaned up